### DISCUSSION TITLE
非同期処理について

### DISCUSSION THREAD

**Discussion by oyui7272** [lit-iphone-course/lit-iphone-course#27] [Question-original] [answered]
### (1) 実現したいこと
Labelに値を入れるupdateLabel関数をAPIの処理(callYOLPAPI関数内のrequest.String)が終わってから呼び出したい。
非同期処理のため、updateLabel関数が先に呼ばれてしまう。

### (2) 実現するために自分で試した内容と結果
callYOLPAPI関数とupdateLabel関数にasyncをつけて、viewDidLoad内でawaitをつけて呼び出したのですが、失敗しました。

### (3) 調べた内容や資料
色々なサイトを参考にしながらasyncやawaitについて調べているのですが、なかなか理解しきれず、解決できていません。
https://techblog.zozo.com/entry/combine-async-await

### (4) 原因だと思われそうな箇所
asyncやawait、Taskの付け型が間違っている...?
asyncを使う場合は必ず返り値がないといけないかもしれない....?　などを考えています

### (5) Githubのリンク
https://github.com/yuipi7272/NanoTour.git

(6) その他
昨日から引き続き聞いてばかりですみません...🙇‍♀️お時間ある時に教えていただけると幸いです。

**Comment by txxxxc** [comment, marked as answer]
やり方としては２つあります！
まず一つ目はcallYOLPAPIの中でUILabelを更新する方法です！761行目のresponse inから始まる部分はクロージャ（簡単に言うと名前のない関数）と言い、Alamofire側がAPIリクエストが終わったタイミングで呼んでくれるやつなのでそこに書くといい感じに更新できると思います！
二つ目はクロージャを渡せる関数を作る方法です！callYOLOAPIがクロージャを引数として受け取れるようにして、response inから始まる所で呼んであげるといけます！
2個目の方法は結構一般的です！

参考はこちらです↓
https://note.com/hoya_pdca/n/n2d54e74219cd

**Reply by txxxxc** [reply]
async awaitの書き方もあるけど結構最近出たやつで知見がネットに転がってないのでクロージャの方がおすすめです！！

**Reply by oyui7272** [reply]
ありがとうございます！！
１つ目の方法、できました！
２つ目の方法でも一応動かすことができたのですが、いまいちあっているか不安なので、ご確認お願いしてもよいでしょうか...??

@escapingをつけないとエラーが出る理由が納得できず...😖
送っていただいたサイトでは、「@escaping をつけていると引数が確定していなくても、関数内のコードを呼び出せる」とありますが、呼出の際に、すでにupdateLabel関数と指定しているのに....?と思い....

-- 修正したところ -- 
ViewDidLoad() {
  callYOLPAPI(completion: updateLabel)
}

callYOLPAPI(completion: @escaping() -> Void) {
  (処理の中身)
  response in {
      completion()
  }
}

**Reply by txxxxc** [reply]
返信遅れました！ごめんなさい🙏
escapingについてはこちらの記事を参照してもらえると良さそうです！
ただ、結構難しい内容ではあるので一旦無視して進めてもらっちゃってOKです！

参考↓
https://qiita.com/mishimay/items/1232dbfe8208e77ed10e

**Reply by yukiny0811** [reply]
escapingが必要なときはXcodeが必ずエラー出してくれるから、最初は無しで書いてみて、escaping必要だよー！ってエラー出されたらそれに従って追加すればOK！

ちゃんと機能まで理解したかったらTTが送ってくれたやつめちゃめちゃわかりやすいね！

**Reply by oyui7272** [reply]
いえいえ！私も返信遅れてしまいすみません！！
分かりました！ありがとうございます！！🙇‍♀️🙏

理解しきれるかはわからないですが、送っていただいたサイト、一度読んでみます！
